dot/rpc/websocketListeners.go

```diff
@@ -0,0 +1,371 @@
+		importedChan:  make(chan *types.Block),
+		wsconn:        c,
+		extrinsic:     types.Extrinsic(extBytes),
+		finalizedChan: make(chan *types.Header),
+	}
+
+	if c.blockAPI == nil {
+		return 0, fmt.Errorf("error BlockAPI not set")
+	}
+	esl.importedChanID, err = c.blockAPI.RegisterImportedChannel(esl.importedChan)
+	if err != nil {
+		return 0, err
+	}
+
+	esl.finalizedChanID, err = c.blockAPI.RegisterFinalizedChannel(esl.finalizedChan)
+	if err != nil {
+		return 0, err
+	}
+
+	c.qtyListeners++
+	esl.subID = c.qtyListeners
+	c.subscriptions[esl.subID] = esl
+	c.blockSubChannels[esl.subID] = esl.importedChanID
+
+	err = c.coreAPI.HandleSubmittedExtrinsic(extBytes)
+	if err != nil {
+		return 0, err
+	}
+	c.safeSend(newSubscriptionResponseJSON(esl.subID, reqID))
+
+	// TODO (ed) since HandleSubmittedExtrinsic has been called we assume the extrinsic is in the tx queue
+	//  should we add a channel to tx queue so we're notified when it's in the queue
+	if c.coreAPI.IsBlockProducer() {
+		c.safeSend(newSubscriptionResponse(AuthorExtrinsicUpdates, esl.subID, "ready"))
+	}
+
+	// todo (ed) determine which peer extrinsic has been broadcast to, and set status
+	return esl.subID, err
+}
+
+// Listen implementation of Listen interface to listen for importedChan changes
+func (l *ExtrinsicSubmitListener) Listen() {
+	// listen for imported blocks with extrinsic
+	go func() {
+		for block := range l.importedChan {
+			if block == nil {
+				continue
+			}
+			exts, err := block.Body.AsExtrinsics()
+			if err != nil {
+				fmt.Printf("error %v\n", err)
+			}
+			for _, v := range exts {
+				if reflect.DeepEqual(v, l.extrinsic) {
+					resM := make(map[string]interface{})
+					resM["inBlock"] = block.Header.Hash().String()
+
+					l.importedHash = block.Header.Hash()
+					l.wsconn.safeSend(newSubscriptionResponse(AuthorExtrinsicUpdates, l.subID, resM))
+				}
+			}
+		}
+	}()
+
+	// listen for finalized headers
+	go func() {
+		for header := range l.finalizedChan {
+			if reflect.DeepEqual(l.importedHash, header.Hash()) {
+				resM := make(map[string]interface{})
+				resM["finalized"] = header.Hash().String()
+				l.wsconn.safeSend(newSubscriptionResponse(AuthorExtrinsicUpdates, l.subID, resM))
+			}
+		}
+	}()
+}
+
+// RuntimeVersionListener to handle listening for Runtime Version
+type RuntimeVersionListener struct {
+	wsconn *WSConn
+	subID  int
+}
+
+func (c *WSConn) initRuntimeVersionListener(reqID float64) (int, error) {
+	rvl := &RuntimeVersionListener{
+		wsconn: c,
+	}
+	if c.coreAPI == nil {
+		c.safeSendError(reqID, nil, "error CoreAPI not set")
+		return 0, fmt.Errorf("error CoreAPI not set")
+	}
+	c.qtyListeners++
+	rvl.subID = c.qtyListeners
+	c.subscriptions[rvl.subID] = rvl
+	initRes := newSubscriptionResponseJSON(rvl.subID, reqID)
+	c.safeSend(initRes)
+
+	return rvl.subID, nil
+}
+
+// Listen implementation of Listen interface to listen for runtime version changes
+func (l *RuntimeVersionListener) Listen() {
```

Comment:
This always sends the runtime version even if the version is not updated. Is this the expected behaviour?

Reply:
Yes, this is to give the subscriber the current value when they first subscribe,  I need to add logic to send updates when the runtime changes.